### ISSUE TITLE
Cannot override yii\web\JqueryAsset for asset/compress

### ISSUE THREAD

**Issue by antongorodezkiy** [yiisoft/yii2#10567]
Hello friends,

I'm trying to compress assets with the following command:

```
./yii asset/compress frontend/config/assets-combine.php frontend/config/assets-prod.php
```

Here is the content of frontend/config/assets-combine.php:

```php
<?php

/**
 * Configuration file for the "yii asset" console command.
 */

// In the console environment, some path aliases may not exist. Please define these:
Yii::setAlias('@webroot', __DIR__ . '/../web');
Yii::setAlias('@web', '/');

return [
    // Adjust command/callback for JavaScript files compressing:
    'jsCompressor' => 'java -jar frontend/nodejs/node_modules/google-closure-compiler/compiler.jar --js {from} --js_output_file {to}',
    // Adjust command/callback for CSS files compressing:
    'cssCompressor' => 'java -jar frontend/nodejs/node_modules/yuicompressor/build/yuicompressor-2.4.8.jar --type css {from} -o {to}',
    // The list of asset bundles to compress:
    'bundles' => [
        'yii\web\JqueryAsset',
        'yii\web\YiiAsset',
        'yii\bootstrap\BootstrapAsset',
        'yii\bootstrap\BootstrapPluginAsset',
		'frontend\assets\ExternalAsset',
		'frontend\assets\BowerAsset',
		'frontend\assets\AppAsset',
    ],
	
    // Asset bundle for compression output:
    'targets' => [
		'frontend.external' => [
			'class' => 'yii\web\AssetBundle',
            'basePath' => '@webroot/assets',
            'baseUrl' => '@web/assets',
            'js' => 'combined/js/external-{hash}.js',
            'css' => 'combined/css/external-{hash}.css',
			'depends' => [
				'yii\web\JqueryAsset',
				'yii\web\YiiAsset',
				'yii\bootstrap\BootstrapAsset',
				'yii\bootstrap\BootstrapPluginAsset',
				'frontend\assets\ExternalAsset',
            ],
		],
		'frontend.vendor' => [
            'class' => 'yii\web\AssetBundle',
            'basePath' => '@webroot/assets',
            'baseUrl' => '@web/assets',
            'js' => 'combined/js/vendor-{hash}.js',
            'css' => 'combined/css/vendor-{hash}.css',
			'depends' => [
				'frontend\assets\BowerAsset'
            ],
        ],
        'frontend.app' => [
            'class' => 'yii\web\AssetBundle',
            'basePath' => '@webroot/assets',
            'baseUrl' => '@web/assets',
            'js' => 'combined/js/app-{hash}.js',
            'css' => 'combined/css/app-{hash}.css',
			'depends' => [
				'frontend\assets\AppAsset',
            ],
        ],
    ],
    
    // Asset manager configuration:
    'assetManager' => [
        'basePath' => '@webroot/assets',
        'baseUrl' => '@web',
        'bundles' => [
            'yii\web\JqueryAsset' => [
               'sourcePath' => null,
               'basePath' => null,
               'js' => ['//cdnjs.cloudflare.com/ajax/libs/jquery/1.11.3/jquery.min.js']
            ]
        ],
        'converter' => [
            'class'=>'nizsheanez\assetConverter\Converter',
            'force' => false, // true : If you want convert your sass each time without time dependency
            'destinationDir' => 'compiled', //at which folder of @webroot put compiled files
            'parsers' => [
                // ...
                'scss' => [ // file extension to parse
                    'class' => 'nizsheanez\assetConverter\Scss',
                    'output' => 'css', // parsed output file type
                    'options' => [ // optional options
                        'enableCompass' => true, // default is true
                        'importPaths' => ['@webroot/scss/'], // import paths, you may use path alias here, 
                            // e.g., `['@path/to/dir', '@path/to/dir1', ...]`
                        'lineComments' => false, // if true — compiler will place line numbers in your compiled output
                        'outputStyle' => 'nested', // May be `compressed`, `crunched`, `expanded` or `nested`,
                            // see more at http://sass-lang.com/documentation/file.SASS_REFERENCE.html#output_style
                    ],
                ],
            ],
        ],
    ],
];

```

Everything works fine except overriding the jQuery asset. This should be a trivial thing to override `yii\web\JqueryAsset` but I spent 2+ hours and still didn't make it work.

In the resulting frontend/config/assets-prod.php I got:

```
[
    'yii\\web\\JqueryAsset' => [
        'sourcePath' => null,
        'js' => [],
        'css' => [],
        'depends' => [],
    ]
]
```

I tried to debug the `AssetController`, `AssetManager` and `AssetBundle`. On this line ( https://github.com/yiisoft/yii2/blob/2.0.6/framework/console/controllers/AssetController.php#L191 ) I have the following object in `$bundles` variable:

```
[yii\web\JqueryAsset] => yii\web\JqueryAsset Object
	(
		[sourcePath] => 
		[js] => Array
			(
				[0] => //cdnjs.cloudflare.com/ajax/libs/jquery/1.11.3/jquery.min.js
			)

		[basePath] => 
		[baseUrl] => 
		[depends] => Array
			(
			)

		[css] => Array
			(
			)

		[jsOptions] => Array
			(
			)

		[cssOptions] => Array
			(
			)

		[publishOptions] => Array
			(
			)

	)
```

And on this line ( https://github.com/yiisoft/yii2/blob/2.0.6/framework/console/controllers/AssetController.php#L192 ) I have the following object in `$targets` variable:

```
[yii\web\JqueryAsset] => yii\web\JqueryAsset Object
	(
		[sourcePath] => /home/.../vendor/bower/jquery/dist
		[js] => Array
			(
				[0] => jquery.js
			)

		[basePath] => 
		[baseUrl] => 
		[depends] => Array
			(
			)

		[css] => Array
			(
			)

		[jsOptions] => Array
			(
			)

		[cssOptions] => Array
			(
			)

		[publishOptions] => Array
			(
			)

	)
```

I believe that's because `Yii::createObject` ( https://github.com/yiisoft/yii2/blob/2.0.6/framework/console/controllers/AssetController.php#L401 ) doesn't care about anything from the $bundles' object.

Maybe this is some kind of bug or I just misunderstood something, but I have no idea how to solve this correct.

**Comment by SilverFire** [comment]
Huge reports are always hard to read.
Could you reduce your report to required minimum for reproducing?

**Comment by antongorodezkiy** [comment]
@SilverFire 

In short:

Yii: 2.0.6

I successfully override `yii\web\JqueryAsset` in application config and used it without compression. When I tried to compress assets with overridden `JqueryAsset` I got the page without jQuery at all (though other custom external/cdn assets were added).

To override jQuery I added following code to the `assetManager` section of `asset/compress` config file.
```
...
'assetManager' => [
        ...
        'bundles' => [
            'yii\web\JqueryAsset' => [
               'sourcePath' => null,
               'basePath' => null,
               'js' => ['//cdnjs.cloudflare.com/ajax/libs/jquery/1.11.3/jquery.min.js']
            ]
        ...
]
```

I made some debug and I think that asset bundle config ignored in the https://github.com/yiisoft/yii2/blob/2.0.6/framework/console/controllers/AssetController.php#L367

**Comment by SilverFire** [comment]
Much easier to read :) I got it, thank you.

**Comment by samdark** [comment]
I guess asset compressor isn't able to get jQuery from that URL.

**Comment by antongorodezkiy** [comment]
@samdark I understand, but I just expected it will work the same way as for other external/cdn assets and it will add script tag with the cdn url.

**Comment by samdark** [comment]
Then you need to exclude it from compression.

**Comment by antongorodezkiy** [comment]
My other assets depends on JqueryAsset. JqueryAsset added in the `depends` section of the bundles. 
So when I tried to compress without JqueryAsset in `depends` section of `targets` in compress config I got the following error:

```
PHP Notice 'yii\base\ErrorException' with message 'Undefined index: yii\web\JqueryAsset'
in /home/.../vendor/yiisoft/yii2/console/controllers/AssetController.php:382
```

I understand that I could remove JqueryAsset dependency from the bundle but what is the reason of `depends` sections in this case? I also could create separate bundles for uncompressed and for compressed versions. But I thought it should "just work™" because other dependency of custom bundle with external assets worked - all external files were added as simple script tags.

**Comment by samdark** [comment]
Could you provide an example of the bundle that worked well?

**Comment by antongorodezkiy** [comment]
```
<?php
/**
 * @link http://www.yiiframework.com/
 * @copyright Copyright (c) 2008 Yii Software LLC
 * @license http://www.yiiframework.com/license/
 */

namespace frontend\assets;

use Yii;
use yii\web\AssetBundle;

/**
 * @author Qiang Xue <qiang.xue@gmail.com>
 * @since 2.0
 */
class ExternalAsset extends AssetBundle
{
    public $basePath = null;
    public $baseUrl = '@web';
    public $css = [
        '//fonts.googleapis.com/css?family=Roboto+Slab',
        '//cdnjs.cloudflare.com/ajax/libs/iCheck/1.0.1/skins/square/_all.css',
        '//cdnjs.cloudflare.com/ajax/libs/magnific-popup.js/0.9.9/magnific-popup.css',
    ];
    public $js = [
        '//cdnjs.cloudflare.com/ajax/libs/lodash.js/3.10.1/lodash.min.js',
        '//cdnjs.cloudflare.com/ajax/libs/moment.js/2.10.6/moment.min.js',
        '//cdnjs.cloudflare.com/ajax/libs/moment-timezone/0.4.0/moment-timezone.min.js',
        '//cdnjs.cloudflare.com/ajax/libs/magnific-popup.js/0.9.9/jquery.magnific-popup.min.js',
        '//cdnjs.cloudflare.com/ajax/libs/iCheck/1.0.1/icheck.min.js',
        '//cdnjs.cloudflare.com/ajax/libs/responsive-nav.js/1.0.32/responsive-nav.min.js',
        '//cdn.datatables.net/1.10.2/js/jquery.dataTables.js'
    ];
    public $depends = [
        'yii\web\JqueryAsset',
    ];
}
```

**Comment by klimov-paul** [comment]
'asset/compress' does not merge overridden asset config with the new created one.
You should re-add your custome JQuery asset config to the one created by command.

**Comment by antongorodezkiy** [comment]
@klimov-paul 

Sorry, not quite understand what do you mean.

If you mean $bundleFile of AssetController::actionCompress(), then I do not understand - this file will be overwritten with the next update:

```
<?php
/**
 * This file is generated by the "yii asset" command.
 * DO NOT MODIFY THIS FILE DIRECTLY.
```

If you mean $configFile of AssetController::actionCompress(), then this is exactly what I'm trying to do ( https://github.com/yiisoft/yii2/issues/10567#issue-126540246 )

**Comment by klimov-paul** [comment]
Provide your `AssetManager` config and the result file of the compress command.

**Comment by antongorodezkiy** [comment]
I added `AssetManager` config for `asset/compress` in the first post https://github.com/yiisoft/yii2/issues/10567#issue-126540246

Result of `asset/compress` command:

```
<?php
/**
 * This file is generated by the "yii asset" command.
 * DO NOT MODIFY THIS FILE DIRECTLY.
 * @version 2016-01-14 15:52:20
 */
return [
    'frontend.external' => [
        'class' => 'yii\\web\\AssetBundle',
        'basePath' => '@webroot/assets',
        'baseUrl' => '@web/assets',
        'js' => [
            'combined/js/external-362e843c2ffa240438454209c7f1d17a.js',
        ],
        'css' => [
            'combined/css/external-6faa0a6858d83c5768842d30347b9951.css',
        ],
    ],
    'frontend.vendor' => [
        'class' => 'yii\\web\\AssetBundle',
        'basePath' => '@webroot/assets',
        'baseUrl' => '@web/assets',
        'js' => [
            'combined/js/vendor-96bbfd79927304335988447a9a90a772.js',
        ],
        'css' => [
            'combined/css/vendor-679c455f0ea0e62d84c858f8f8eb74a5.css',
        ],
    ],
    'frontend.app' => [
        'class' => 'yii\\web\\AssetBundle',
        'basePath' => '@webroot/assets',
        'baseUrl' => '@web/assets',
        'js' => [
            'combined/js/app-32d531296a9b4a59cbb1b2ac892eacce.js',
        ],
        'css' => [
            'combined/css/app-3755c49c09ca36eee5aeab46893b3e3b.css',
        ],
    ],
    'yii\\web\\JqueryAsset' => [
        'sourcePath' => null,
        'js' => [],
        'css' => [],
        'depends' => [],
    ],
    'yii\\web\\YiiAsset' => [
        'sourcePath' => null,
        'js' => [],
        'css' => [],
        'depends' => [
            'yii\\web\\JqueryAsset',
            'frontend.external',
        ],
    ],
    'yii\\bootstrap\\BootstrapAsset' => [
        'sourcePath' => null,
        'js' => [],
        'css' => [],
        'depends' => [
            'frontend.external',
        ],
    ],
    'yii\\bootstrap\\BootstrapPluginAsset' => [
        'sourcePath' => null,
        'js' => [],
        'css' => [],
        'depends' => [
            'yii\\web\\JqueryAsset',
            'yii\\bootstrap\\BootstrapAsset',
            'frontend.external',
        ],
    ],
    'frontend\\assets\\ExternalAsset' => [
        'basePath' => null,
        'baseUrl' => '',
        'css' => [
            '//fonts.googleapis.com/css?family=Roboto+Slab',
            '//cdnjs.cloudflare.com/ajax/libs/iCheck/1.0.1/skins/square/_all.css',
            '//cdnjs.cloudflare.com/ajax/libs/magnific-popup.js/0.9.9/magnific-popup.css',
        ],
        'js' => [
            '//cdnjs.cloudflare.com/ajax/libs/lodash.js/3.10.1/lodash.min.js',
            '//cdnjs.cloudflare.com/ajax/libs/moment.js/2.10.6/moment.min.js',
            '//cdnjs.cloudflare.com/ajax/libs/moment-timezone/0.4.0/moment-timezone.min.js',
            '//cdnjs.cloudflare.com/ajax/libs/magnific-popup.js/0.9.9/jquery.magnific-popup.min.js',
            '//cdnjs.cloudflare.com/ajax/libs/iCheck/1.0.1/icheck.min.js',
            '//cdnjs.cloudflare.com/ajax/libs/responsive-nav.js/1.0.32/responsive-nav.min.js',
            '//cdn.datatables.net/1.10.2/js/jquery.dataTables.js',
            'site/translations-js',
        ],
        'depends' => [
            'yii\\web\\JqueryAsset',
        ],
        'sourcePath' => null,
        'jsOptions' => [],
        'cssOptions' => [],
        'publishOptions' => [],
        'class' => 'frontend\\assets\\ExternalAsset',
    ],
    'frontend\\assets\\BowerAsset' => [
        'sourcePath' => null,
        'js' => [],
        'css' => [],
        'depends' => [
            'yii\\web\\JqueryAsset',
            'yii\\web\\YiiAsset',
            'yii\\bootstrap\\BootstrapAsset',
            'yii\\bootstrap\\BootstrapPluginAsset',
            'frontend\\assets\\ExternalAsset',
            'frontend.vendor',
        ],
    ],
    'frontend\\assets\\AppAsset' => [
        'sourcePath' => null,
        'js' => [],
        'css' => [],
        'depends' => [
            'frontend\\assets\\BowerAsset',
            'frontend.app',
        ],
    ],
];
```

**Comment by klimov-paul** [comment]
Just as I said.
The source `JqueryAsset`:

```php
            'yii\web\JqueryAsset' => [
               'sourcePath' => null,
               'basePath' => null,
               'js' => ['//cdnjs.cloudflare.com/ajax/libs/jquery/1.11.3/jquery.min.js']
            ]
```

compressed `JqueryAsset`:

```php
'yii\\web\\JqueryAsset' => [
        'sourcePath' => null,
        'js' => [],
        'css' => [],
        'depends' => [],
    ],
```

Thus you need to merge compression result with your custom config.

**Comment by klimov-paul** [comment]
Perhaps it is possible to adjust command in a way, it can handle such situation on its own. We can try to copy file lists (`js` and `css`) in case of external asset bundle instead of ignoring them.

See `buildTarget()` method:

https://github.com/yiisoft/yii2/blob/master/framework/console/controllers/AssetController.php#L347

**Comment by samdark** [comment]
I think @antongorodezkiy wants to ignore these.

**Comment by klimov-paul** [comment]
> I think @antongorodezkiy wants to ignore these.

Unlikely. The problem is he overrides `JqueryAsset` in the way it becomes external istead of published one. While being compressed it is processed as extrernal: skipped without adjusting, which means its `js` property is set to empty array. For normal bundle this will cause NO changes of original(!) `js` value.
`JqueryAsset` could load in normal way (being published), but override prevents it, because it set `sourcePath` to `null`. Thus there is no publish of orignal js and there is no extrenal link.

**Comment by antongorodezkiy** [comment]
> Thus you need to merge compression result with your custom config.

Yes, in that way (when added manually) it works fine. I thought that this is some kind of bug.

**Comment by klimov-paul** [comment]
Resolved by commit db15c505579501a5996ff15f46ebecf52c8e8755
